spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/TaskLifecycleListener.java

```diff
@@ -169,5 +179,42 @@ private void doTaskStart() {
 			logger.error("Multiple start events have been received.  The first one was " +
 					"recorded.");
 		}
+		invokeOnTaskStartup(taskExecution);
+	}
+
+	private void invokeOnTaskStartup(TaskExecution taskExecution){
+		if (taskExecutionListeners != null) {
+			for (TaskExecutionListener taskExecutionListener : taskExecutionListeners) {
+				taskExecutionListener.onTaskStartup(getTaskExecutionCopy(taskExecution));
+			}
+		}
+	}
+
+	private void invokeOnTaskEnd(TaskExecution taskExecution){
+		if (taskExecutionListeners != null) {
+			for (TaskExecutionListener taskExecutionListener : taskExecutionListeners) {
+				taskExecutionListener.onTaskEnd(getTaskExecutionCopy(taskExecution));
+			}
+		}
+	}
+
+	private void invokeOnTaskError(TaskExecution taskExecution, Throwable throwable){
+		if (taskExecutionListeners != null) {
+			for (TaskExecutionListener taskExecutionListener : taskExecutionListeners) {
+				taskExecutionListener.onTaskFailed(getTaskExecutionCopy(taskExecution),
+						throwable);
+			}
+		}
+	}
```

Comment:
So it's good that we're providing a copy of the `TaskExecution`.  I'm wondering if we should allow anything to be copied back (edit message is one thing I'm wondering)...